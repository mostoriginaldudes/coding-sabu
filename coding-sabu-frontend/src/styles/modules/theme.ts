import { css } from '@emotion/react';
import { CSSInterpolation } from '@emotion/serialize';
import openColor from 'open-color';

interface Sizes {
  desktop: 1024;
  tablet: 768;
  unitBig: 64;
  unitRegular: 48;
}

const sizes: Sizes = {
  desktop: 1024,
  tablet: 768,
  unitBig: 64,
  unitRegular: 48
};

const media = {
  desktop: (...args: TemplateStringsArray[]) =>
    css`
      @media (max-width: ${sizes.desktop / 16}em) {
        ${css(...(args as unknown as CSSInterpolation[]))}
      }
    `,
  tablet: (...args: TemplateStringsArray[]) =>
    css`
      @media (max-width: ${sizes.tablet / 16}em) {
        ${css(...(args as unknown as CSSInterpolation[]))}
      }
    `
} as const;

const colors = openColor;

export { sizes, media, colors };
