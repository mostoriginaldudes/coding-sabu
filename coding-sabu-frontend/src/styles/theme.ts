import { css } from "@emotion/react";
import openColor from "open-color";

interface Sizes {
  desktop: 1024;
  tablet: 768;
  unitDesktop: "64px";
  unitTablet: "48px";
}

const sizes: Sizes = {
  desktop: 1024,
  tablet: 768,
  unitDesktop: "64px",
  unitTablet: "48px"
};

const media = {
  desktop: (...args: any) =>
    css`
      @media (max-width: ${sizes.desktop / 16}em) {
        ${css(...args)}
      }
    `,
  tablet: (...args: any) =>
    css`
      @media (max-width: ${sizes.tablet / 16}em) {
        ${css(...args)}
      }
    `
} as const;

const colors = openColor;

export { sizes, media, colors };
