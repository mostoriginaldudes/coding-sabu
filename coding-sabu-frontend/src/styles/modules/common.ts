import styled from '@emotion/styled';
import { css } from '@emotion/react';
import { media } from './theme';

export const flexCenter = css`
  display: flex;
  justify-content: center;
  align-items: center;
`;

export const positionFixed = css`
  position: fixed;
  top: 0;
  left: 0;
`;

export const FlexRow = styled.div`
  ${flexCenter}
`;

export const FlexCol = styled.div`
  ${flexCenter}
  flex-direction: column;
`;

export const Row = styled.div`
  ${flexCenter}
  justify-content: space-between;
  width: 100%;
  margin: 1em 0;

  ${media.tablet`
    flex-direction: column;
  `}

  & > div {
    width: 100%;
  }

  & > button {
    width: 45%;

    ${media.tablet`
      width: 100%;
      margin: 5px 0;
    `}
  }
`;
