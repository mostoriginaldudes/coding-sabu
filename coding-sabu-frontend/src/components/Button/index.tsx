import styled from '@emotion/styled';
import { media, colors } from '../../styles/theme';

interface ButtonProps {
  color: 'yellow' | 'black' | 'white';
  radius: number;
}

const Button = styled.button<ButtonProps>`
  border-radius: ${({ radius }) => radius}px;

  border: ${({ color }) =>
    `1px solid ${color === 'yellow' ? colors.yellow[4] : colors.black}`};

  background-color: ${({ color }) =>
    color === 'yellow' ? colors.yellow[4] : colors[color]};

  color: ${({ color }) => (color === 'black' ? colors.white : colors.black)};
  padding: 5px 7px;
  margin-left: 10px;
  font-size: 0.8em;
  cursor: pointer;
  ${media.tablet`
    display: none;
  `}
`;

export default Button;
