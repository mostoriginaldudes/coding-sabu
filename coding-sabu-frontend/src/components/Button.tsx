import styled from '@emotion/styled';
import { ButtonHTMLAttributes, ReactNode } from 'react';
import { colors } from 'styles/modules/theme';
const { yellow, black, white } = colors;

interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  color: 'yellow' | 'black' | 'white';
  children: ReactNode | string;
  radius?: number;
  height?: number;
}

const Button = styled.button<ButtonProps>`
  height: ${({ height }) => height}rem;
  border-radius: ${({ radius }) => radius}px;
  border: ${({ color }) => `1px solid ${color === 'yellow' ? yellow[4] : black}`};
  background-color: ${({ color }) => (color === 'yellow' ? yellow[4] : colors[color])};
  color: ${({ color }) => (color === 'black' ? white : black)};
  padding: 5px 7px;
  font-size: 0.8em;
  cursor: pointer;
  box-shadow: 1px 12px 15px -4px rgba(0, 0, 0, 0.62);

  &:hover {
    box-shadow: 1px 12px 15px 0px rgba(0, 0, 0, 0.62);
    transition: box-shadow 250ms ease;
  }

  &:active {
    box-shadow: 1px 12px 15px -4px rgba(0, 0, 0, 0.3);
    transition: box-shadow 200ms ease;
  }

  & > a {
    color: ${({ color }) => (color === 'black' ? white : black)};
  }
`;

Button.defaultProps = {
  color: 'black',
  radius: 0
} as ButtonProps;

export default Button;
