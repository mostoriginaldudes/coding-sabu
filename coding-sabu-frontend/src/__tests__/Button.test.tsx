import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { matchers } from '@emotion/jest';
import Button from 'components/Button';
import { colors } from 'styles/modules/theme';

expect.extend(matchers);

const onClick = jest.fn();

const renderButton = (color: 'yellow' | 'white' | 'black') => {
  render(
    <Button color={color} onClick={onClick}>
      버튼
    </Button>
  );
  return screen.getByRole('button');
};

describe('<Button /> UI 테스트', () => {
  describe('color prop이 잘 전달된다.', () => {
    test('버튼 색깔은 노란색이다.', () => {
      const ButtonComponent = renderButton('yellow');
      expect(ButtonComponent).toHaveStyleRule('background-color', colors.yellow[4]);
    });

    test('버튼 색깔은 검은색이다.', () => {
      const ButtonComponent = renderButton('black');
      expect(ButtonComponent).toHaveStyleRule('background-color', colors.black);
    });

    test('버튼 색깔은 하얀색이다.', () => {
      const ButtonComponent = renderButton('white');
      expect(ButtonComponent).toHaveStyleRule('background-color', colors.white);
    });
  });

  describe('props 잘 전달된다.', () => {
    test('버튼 글자 보인다.', () => {
      const ButtonComponent = renderButton('white');
      expect(ButtonComponent).toHaveTextContent('버튼');
    });

    test('버튼이 한번 클릭된다.', () => {
      const ButtonComponent = renderButton('white');
      userEvent.click(ButtonComponent);
      expect(onClick.mock.calls.length).toBe(1);
    });

    test('onClick으로 전달된 함수가 호출된다.', () => {
      const ButtonComponent = renderButton('white');
      userEvent.click(ButtonComponent);
      expect(onClick).toBeCalled();
    });
  });
});
