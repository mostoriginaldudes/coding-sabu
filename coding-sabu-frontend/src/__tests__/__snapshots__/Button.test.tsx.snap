// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> UI 테스트 버튼 색깔은 검정 1`] = `
.emotion-0 {
  height: rem;
  border-radius: 0px;
  border: 1px solid #000000;
  background-color: #000000;
  color: #ffffff;
  padding: 5px 7px;
  font-size: 0.8em;
  cursor: pointer;
  box-shadow: 1px 12px 15px -4px rgba(0, 0, 0, 0.62);
}

.emotion-0:hover {
  box-shadow: 1px 12px 15px 0px rgba(0, 0, 0, 0.62);
  -webkit-transition: box-shadow 250ms ease;
  transition: box-shadow 250ms ease;
}

.emotion-0:active {
  box-shadow: 1px 12px 15px -4px rgba(0, 0, 0, 0.3);
  -webkit-transition: box-shadow 200ms ease;
  transition: box-shadow 200ms ease;
}

.emotion-0>a {
  color: #ffffff;
}

<button
  class="emotion-0"
  color="black"
  radius="0"
>
  버튼
</button>
`;

exports[`<Button /> UI 테스트 버튼 색깔은 노랑 1`] = `
.emotion-0 {
  height: rem;
  border-radius: 0px;
  border: 1px solid #ffd43b;
  background-color: #ffd43b;
  color: #000000;
  padding: 5px 7px;
  font-size: 0.8em;
  cursor: pointer;
  box-shadow: 1px 12px 15px -4px rgba(0, 0, 0, 0.62);
}

.emotion-0:hover {
  box-shadow: 1px 12px 15px 0px rgba(0, 0, 0, 0.62);
  -webkit-transition: box-shadow 250ms ease;
  transition: box-shadow 250ms ease;
}

.emotion-0:active {
  box-shadow: 1px 12px 15px -4px rgba(0, 0, 0, 0.3);
  -webkit-transition: box-shadow 200ms ease;
  transition: box-shadow 200ms ease;
}

.emotion-0>a {
  color: #000000;
}

<button
  class="emotion-0"
  color="yellow"
  radius="0"
>
  버튼
</button>
`;

exports[`<Button /> UI 테스트 버튼 색깔은 하양 1`] = `
.emotion-0 {
  height: rem;
  border-radius: 0px;
  border: 1px solid #000000;
  background-color: #ffffff;
  color: #000000;
  padding: 5px 7px;
  font-size: 0.8em;
  cursor: pointer;
  box-shadow: 1px 12px 15px -4px rgba(0, 0, 0, 0.62);
}

.emotion-0:hover {
  box-shadow: 1px 12px 15px 0px rgba(0, 0, 0, 0.62);
  -webkit-transition: box-shadow 250ms ease;
  transition: box-shadow 250ms ease;
}

.emotion-0:active {
  box-shadow: 1px 12px 15px -4px rgba(0, 0, 0, 0.3);
  -webkit-transition: box-shadow 200ms ease;
  transition: box-shadow 200ms ease;
}

.emotion-0>a {
  color: #000000;
}

<button
  class="emotion-0"
  color="white"
  radius="0"
>
  버튼
</button>
`;
